In [1]:
import pandas as pd
import pandas_datareader as web
import seaborn as sns
from datetime import datetime

In [2]:
f = web.DataReader('USD000UTSTOM', 'moex', start='2017-07-01', end='2017-07-31')
f.head()

BOARDID   SHORTNAME         SECID   OPEN     LOW     HIGH    CLOSE  \
TRADEDATE                                                                       
2017-07-03    CNGD  USDRUB_TOM  USD000UTSTOM  58.98  58.840  59.4250  59.3600   
2017-07-04    CETS  USDRUB_TOM  USD000UTSTOM  59.30  59.135  59.4575  59.4125   
2017-07-04    CNGD  USDRUB_TOM  USD000UTSTOM  59.36  58.930  59.3600  59.3575   
2017-07-05    CETS  USDRUB_TOM  USD000UTSTOM  59.30  59.300  60.2600  59.9825   
2017-07-05    CNGD  USDRUB_TOM  USD000UTSTOM  59.34  59.265  60.1800  60.1800   

            NUMTRADES        VOLRUR  WAPRICE  
TRADEDATE                                     
2017-07-03         24  1.864785e+09      NaN  
2017-07-04      21053  1.090265e+11  59.2700  
2017-07-04         37  1.046416e+09      NaN  
2017-07-05      50108  2.874226e+11  59.9234  
2017-07-05         35  6.339036e+09      NaN

In [3]:
sber=web.DataReader('SBER', 'moex', start='2017-01-01', end='2017-12-31')
sber.head(5)

BOARDID SHORTNAME SECID  NUMTRADES         VALUE    OPEN       LOW  \
TRADEDATE                                                                       
2017-01-04    RPMA  Сбербанк  SBER        211  8.429459e+08  180.12  140.8842   
2017-01-05    RPMA  Сбербанк  SBER        222  8.436439e+08  175.95  140.8842   
2017-01-06    RPMA  Сбербанк  SBER        208  8.423766e+08  171.90  140.8842   
2017-01-09    RPMA  Сбербанк  SBER        213  8.618491e+08  170.69  142.6355   
2017-01-10    RPMA  Сбербанк  SBER        209  8.516869e+08  168.35  141.0087   

               HIGH   CLOSE  WAPRICE  
TRADEDATE                             
2017-01-04  186.714  169.74      NaN  
2017-01-05  186.714  169.74      NaN  
2017-01-06  186.714  169.74      NaN  
2017-01-09  189.035  171.85      NaN  
2017-01-10  186.879  171.60      NaN

In [4]:
from finam.export import Exporter, Market, Timeframe
exporter = Exporter()

In [5]:
exporter.lookup(code='LKOH')

name  code  market
id                                      
8                   ЛУКОЙЛ  LKOH       1
332                   LKOH  LKOH       3
918                 Лукойл  LKOH       8
17453                 LKOH  LKOH      14
22427   LKOH(RTS-Standard)  LKOH      38
8                   ЛУКОЙЛ  LKOH     200
419710  ПАО \'ЛУКОЙЛ\', ао  LKOH     517

Коды рынков на экспорт (не все, остальные смотрите на finam.ru, но они не качаютя)    
    BONDS = 2<br>
    COMMODITIES = 24<br>
    CURRENCIES = 45<br>
    ETF = 28<br>
    ETF_MOEX = 515<br>
    FUTURES = 14  # non-expired futures<br>
    FUTURES_ARCHIVE = 17  # expired futures<br>
    FUTURES_USA = 7<br>
    INDEXES = 6<br>
    SHARES = 1<br>
    SPB = 517<br>
    USA = 25<br>

In [6]:
# посмотрим что есть на рынках ММВБ
market = exporter.lookup(market=Market.SHARES)
display(market.head())
print(','.join(market['code']))

name   code  market
id                               
436091  СибГост ао   SIBG       1
81940   Химпром ап  HIMCP       1
19095       ЦМТ ао   WTCM       1
81943    КУЗОЦМ ао   KUNF       1
19096       ЦМТ ап  WTCMP       1

SIBG,HIMCP,WTCM,KUNF,WTCMP,MFON,KTSBP,HIMC,KTSB,STSB,TCSG,RTKM,RTKMP,AVAN,LNTA,STSBP,ENPL,SNGS,RSTIP,PHOR,UCSS,OMZZP,OBUV,SAGOP,SAGO,SNGSP,LNZL,TANLP,TANL,KCHE,KLSB,FESH,RSTI,KCHEP,LSNGP,LSNG,VDSB,KZOSP,KZMS,VZRZP,KRKOP,VGSB,VGSBP,URKZ,ROST,AFLT,CHGZ,AMEZ,PRFN,TUCH,AGRO,MVID,VZRZ,KMEZ,KRKO,VJGZ,VJGZP,MGNT,MAGEP,MAGE,MTSS,TRNFP,FTRE,BSPB,TUZA,RAVN,FIVE,RUAL,DVEC,DASB,LNZLP,LIFE,MFGS,MFGSP,KRSB,NSVZ,KGKCP,MSTT,MOBB,MOEX,KRSBP,HALS,SLEN,ROSB,RLMNP,RLMN,NFAZ,MRSB,MORI,KUBE,KGKC,MSST,KUZB,BISV,ROSN,BISVP,SIBN,USBN,ACKO,NKNC,NKNCP,NKHP,KAZTP,KBTK,URKA,NLMK,NMTP,RTGZ,CHKZ,KROT,FEES,KRKN,LSRG,LKOH,LPSB,BANEP,NPOF,KRKNP,KSGR,TGKDP,TGKD,UWGN,TGKA,ODVA,OGKB,LVHK,KOGK,KMTZ,YNDX,KROTP,NKSH,NVTK,TGKB,MTLRP,MTLR,ABRD,BANE,TGKN,SAREP,SARE,RASP,GTRK,ALNU,ALRS,TGKBP,DSKY,KZOS,CHMK,ENRU,PRTK,SVAV,RKKE,PAZA,PMSBP,PMSB,VSYDP,CNTLP,CNTL,RZSB,SBER,PLSM,UTAR,MAGN,UNKL,DIOD,SBERP,VSYD,VTBR,RUSI,BLNG,BELU,NNSBP,RTSB,RTSBP,YRSB,ZILL,MSRS,PIKK,TNSE,MGTSP,ZVEZ,DZRDP,DZRD,ALBK,YRSBP,NNSB,UNAC,AKRN,CHMF,RBCM,MGNZ,YK

In [7]:
# посмотрим что есть на товарных рынках
market = exporter.lookup(market=Market.COMMODITIES)
display(market.head())
print(','.join(market['code']))

name          code  market
id                                        
485787  Brent(Apr22)  Brent.22.Apr      24
498510  Brent(Jul23)  Brent.23.Jul      24
485765  Brent(Jun20)  Brent.20.Jun      24
485761  Brent(Feb20)  Brent.20.Feb      24
485763  Brent(Apr20)  Brent.20.Apr      24

Brent.22.Apr,Brent.23.Jul,Brent.20.Jun,Brent.20.Feb,Brent.20.Apr,Brent.24.Jul,Brent.23.Apr,PA,Brent.21.Aug,Brent.21.Apr,Brent.20.Aug,Light.22.Dec,Brent.21.Feb,Brent.24.Apr,Brent.23.Aug,BZ,NG,Brent.24.Jan,Light.21.Oct,Light.20.Nov,Light.21.Nov,Brent.21.Jun,Light.22.Nov,Brent.21.Dec,Light.23.Nov,Light.24.May,Brent.23.Jan,Brent.20.Dec,Brent.20.Jan,Brent.22.Jan,Brent.19.Dec,Brent.21.Jan,Light.20.Oct,Brent.22.Jul,Brent.22.Dec,Light.23.Feb,Brent.22.Aug,Brent.23.Dec,Brent.22.Feb,Brent.23.Feb,Brent.21.Jul,Brent.24.Feb,Brent.20.Jul,Brent.24.Aug,Light.21.May,Light.20.Feb,Light.21.Feb,Light.22.May,Light.23.Oct,Light.22.Feb,Light.23.May,Light.22.Oct,Light.24.Oct,Brent.22.Jun,Light.20.Apr,Light.20.Aug,Light.24.Jun,Light.24.Feb,Brent.20.Sep,Light.22.Jan,ZW,Light.22.Jul,Light.21.Jul,GC,Light.23.Jun,Light.23.Apr,Light.23.Jul,Brent.23.Nov,Light.22.Aug,RB,Brent.22.Nov,Brent.21.Nov,Light.20.Mar,Brent.20.Nov,Light.23.Jan,Light.21.Mar,Light.20.May,Light.22.Jun,Light.23.Dec,Brent.24.Oct,Brent.24.Sep,Brent.2

In [8]:
# тикер BRENT
market = exporter.lookup(code = "BZ", market=Market.COMMODITIES)
display(market.head())

name code  market
id                       
19473  Brent   BZ      24

In [9]:
start_date = datetime(2017, 1, 1)
end_date = datetime(2017, 12, 31)
sber = exporter.lookup(code='SBER',market=Market.SHARES)

display(sber)

data_daily = exporter.download(
    sber.index[0], 
    market=Market.SHARES,
    start_date=start_date,
    end_date=end_date
)
data_daily.head()


name  code  market
id                        
3   Сбербанк  SBER       1

<OPEN>  <HIGH>   <LOW>  <CLOSE>     <VOL>
index                                                
2017-01-03  173.41  181.68  173.00   180.12  53842180
2017-01-04  178.91  179.41  174.23   175.95  41327620
2017-01-05  176.32  177.28  171.54   171.90  35076900
2017-01-06  172.41  173.75  169.50   170.69  35395940
2017-01-09  171.37  172.39  168.05   168.35  35166440

In [10]:
#Timeframe:
# TICKS
# MINUTES1
# MINUTES5
# MINUTES10
# MINUTES15
# MINUTES30
# HOURLY
# DAILY
# WEEKLY
# MONTHLY

data_5min = exporter.download(
    id_ = sber.index[0],
    market = Market.SHARES,
    start_date = datetime(2017, 1, 1),
    end_date = datetime(2017, 1, 5),
    timeframe = Timeframe.MINUTES5 #
)
data.head(10)

NameError: name 'data' is not defined

Поработаем с дневными данными

In [ ]:
data_daily.head(5)

In [12]:
data_daily.columns = ['Open', 'High','Low','Close','Volume']
data_daily.head(5)

Open    High     Low   Close    Volume
index                                               
2017-01-03  173.41  181.68  173.00  180.12  53842180
2017-01-04  178.91  179.41  174.23  175.95  41327620
2017-01-05  176.32  177.28  171.54  171.90  35076900
2017-01-06  172.41  173.75  169.50  170.69  35395940
2017-01-09  171.37  172.39  168.05  168.35  35166440

In [13]:
data_volume_month=data_daily.Volume.resample('m').sum()
data_volume_month

index
2017-01-31     988723570
2017-02-28     816284690
2017-03-31     978856860
2017-04-30     963970030
2017-05-31     824194850
2017-06-30    1247585930
2017-07-31    1055435570
2017-08-31    1063813850
2017-09-30     942566460
2017-10-31     744449580
2017-11-30    1252529590
2017-12-31     682568440
Freq: M, Name: Volume, dtype: int64

In [14]:
open_m=data_daily.Open.resample('M').first().values
open_m

array([173.41, 172.51, 155.99, 160.5 , 166.  , 156.4 , 146.69, 165.53,
       183.85, 192.71, 194.5 , 224.53])

In [15]:
close_m=data_daily.Close.resample('M').last().values
low_m=data_daily.Low.resample('M').min().values
high_m=data_daily.High.resample('M').max().values

In [16]:
month=data_daily.index.strftime('%b').unique()
month

Index(['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct',
       'Nov', 'Dec'],
      dtype='object')

In [17]:
data_daily.to_period('M').index.unique()

PeriodIndex(['2017-01', '2017-02', '2017-03', '2017-04', '2017-05', '2017-06',
             '2017-07', '2017-08', '2017-09', '2017-10', '2017-11', '2017-12'],
            dtype='period[M]', name='index', freq='M')

In [18]:
moth_a = pd.DataFrame(
    {'month':month,'open':open_m,'close':close_m,'low':low_m,'high':high_m},
    index=data_daily.to_period('M').index.unique())
moth_a.head(3)

month    open  close     low    high
index                                       
2017-01   Jan  173.41  172.2  163.34  181.68
2017-02   Feb  172.51  156.0  155.70  174.98
2017-03   Mar  155.99  159.8  155.23  165.99

In [19]:
sb_m = data_daily.resample('m').agg({'Open':'first','Low':'min','High':'max','Close':'last'})
sb_m

Open     Low    High   Close
index                                     
2017-01-31  173.41  163.34  181.68  172.20
2017-02-28  172.51  155.70  174.98  156.00
2017-03-31  155.99  155.23  165.99  159.80
2017-04-30  160.50  147.62  168.45  165.20
2017-05-31  166.00  155.03  172.39  155.93
2017-06-30  156.40  136.20  158.89  145.59
2017-07-31  146.69  146.22  167.35  164.53
2017-08-31  165.53  165.30  186.33  183.51
2017-09-30  183.85  180.93  195.00  192.33
2017-10-31  192.71  189.06  198.52  193.80
2017-11-30  194.50  192.38  233.95  224.35
2017-12-31  224.53  219.23  230.50  225.20

In [20]:
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.graph_objs as go
init_notebook_mode(connected=True)

In [25]:
trace = go.Ohlc(x=data_daily.index,
                open=data_daily['Open'],
                high=data_daily['High'],
                low=data_daily['Low'],
                close=data_daily['Close'])

data = [trace]
iplot(data)

In [24]:
trace2 = go.Candlestick(
                x=sb_m.index,# Проблема месяц, есть ли решение на середину месяца
                open=sb_m['Open'],
                high=sb_m['High'],
                low=sb_m['Low'],
                close=sb_m['Close'])


layout = go.Layout(
    xaxis = dict(
        rangeslider = dict(
            visible = False
        )
    )
)

data2 = [trace2]
fig = go.Figure(data=data2,layout=layout)
iplot(fig)

In [26]:
month=sb_m.index.strftime('%b-2017')

trace2 = go.Candlestick(
                x=month, # Внимание на месяц
                open=sb_m['Open'],
                high=sb_m['High'],
                low=sb_m['Low'],
                close=sb_m['Close'])


# layout = go.Layout(
#     xaxis = dict(
#         rangeslider = dict(
#             visible = False
#         )
#     )
# )

data2 = [trace2]
fig = go.Figure(data=data2)#,layout=layout)
iplot(fig)